libc/src/__support/OSUtil/linux/random.cpp

```diff
@@ -0,0 +1,359 @@
+    if (new_allocation == MAP_FAILED)
+      return false;
+
+    for (size_t i = 0; i < pages_per_allocation; ++i) {
+      auto *page = new_allocation + i * page_size;
+      for (size_t j = 0; j < states_per_page; ++j)
+        freelist.push_unchecked(page + j * params.size_of_opaque_state);
+    }
+    return true;
+  }
+
+  static RandomStateFactory *instance();
+
+  void *acquire() {
+    cpp::lock_guard guard{mutex};
+    if (freelist.empty() && !allocate_new_states())
+      return nullptr;
+    return freelist.pop();
+  }
+  void release(void *state) {
+    cpp::lock_guard guard{mutex};
+    // there should be no need to check this pushing
+    freelist.push_unchecked(state);
+  }
+  ~RandomStateFactory() {
+    for (auto *allocation : allocations)
+      munmap(allocation, page_size * pages_per_allocation);
+  }
+
+public:
+  static void *acquire_global() {
+    auto *factory = instance();
+    if (!factory)
+      return nullptr;
+    return factory->acquire();
+  }
+  static void release_global(void *state) {
+    auto *factory = instance();
+    if (!factory)
+      return;
+    factory->release(state);
+  }
+  static size_t size_of_opaque_state() {
+    return instance()->params.size_of_opaque_state;
+  }
+  static void postfork_cleanup();
+};
+
+thread_local bool fork_inflight = false;
+thread_local void *tls_state = nullptr;
+alignas(RandomStateFactory) static char factory_storage[sizeof(
+    RandomStateFactory)]{};
+static CallOnceFlag factory_onceflag = callonce_impl::NOT_CALLED;
+static bool factory_valid = false;
+
+RandomStateFactory *RandomStateFactory::instance() {
+  callonce(&factory_onceflag, []() {
+    auto *factory = new (factory_storage) RandomStateFactory();
+    factory_valid = factory->prepare();
+    if (factory_valid)
+      atexit([]() {
+        auto factory = reinterpret_cast<RandomStateFactory *>(factory_storage);
+        factory->~RandomStateFactory();
+        factory_valid = false;
+      });
+  });
+  return factory_valid ? reinterpret_cast<RandomStateFactory *>(factory_storage)
+                       : nullptr;
+}
+
+void RandomStateFactory::postfork_cleanup() {
+  if (factory_valid)
+    reinterpret_cast<RandomStateFactory *>(factory_storage)
+        ->~RandomStateFactory();
+  factory_onceflag = callonce_impl::NOT_CALLED;
+  factory_valid = false;
+}
+
+void *acquire_tls() {
+  if (fork_inflight)
+    return nullptr;
+  // previous acquire failed, do not try again
+  if (tls_state == MAP_FAILED)
+    return nullptr;
+  // first acquirement
+  if (tls_state == nullptr) {
+    tls_state = RandomStateFactory::acquire_global();
+    // if still fails, remember the failure
+    if (tls_state == nullptr) {
+      tls_state = MAP_FAILED;
+      return nullptr;
+    } else {
+      // register the release callback.
+      if (__cxa_thread_atexit_impl(
+              [](void *s) { RandomStateFactory::release_global(s); }, tls_state,
```

Comment:
What happens if random_fill is called during or after the atexit callback?

Reply:
We will left it as UB in the latest implementation.